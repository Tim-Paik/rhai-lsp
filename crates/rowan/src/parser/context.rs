//! The parser context is a separate module to limit
//! the API surface for the parser functions.
#![allow(dead_code)]
use rowan::{GreenNodeBuilder, TextRange, TextSize};

use crate::syntax::{
    Lexer,
    SyntaxKind::{self, *},
};

use super::{Parse, ParseError, ParseErrorKind};

pub(crate) struct Context<'src> {
    lexer: Lexer<'src>,
    current_token: Option<SyntaxKind>,
    last_token: Option<SyntaxKind>,
    green: GreenNodeBuilder<'static>,
    errors: Vec<ParseError>,

    /// Tracks statements being separated by ";".
    statement_closed: bool,
}

impl<'src> Context<'src> {
    pub(crate) fn new(source: &'src str) -> Self {
        Self {
            lexer: Lexer::new(source),
            current_token: None,
            last_token: None,
            green: GreenNodeBuilder::new(),
            errors: Vec::new(),

            statement_closed: true,
        }
    }

    pub(crate) fn token(&mut self) -> Option<SyntaxKind> {
        // Eat insignificant tokens
        loop {
            if self.current_token.is_none() {
                self.current_token = self.lexer.next();
            }

            match self.current_token {
                Some(COMMENT_BLOCK) | Some(COMMENT_LINE) | Some(WHITESPACE) => {
                    self.eat();
                    self.last_token = None // We aren't interested in these.
                }
                Some(ERROR) => {
                    self.eat_error(ParseErrorKind::InvalidInput);
                    self.last_token = None // We aren't interested in these.
                }
                _ => break,
            }
        }

        self.current_token
    }

    pub(crate) fn previous_token(&self) -> Option<SyntaxKind> {
        self.last_token
    }

    pub(crate) fn eat(&mut self) {
        if let Some(t) = self.current_token.take() {
            self.green.token(t.into(), self.lexer.slice());
            self.last_token = Some(t);
        }
        self.current_token = None;
    }

    pub(crate) fn eat_as(&mut self, kind: SyntaxKind) {
        if self.current_token.is_some() {
            self.green.token(kind.into(), self.lexer.slice());
            self.last_token = Some(kind);
        }
        self.current_token = None;
    }

    pub(crate) fn eat_error(&mut self, error: ParseErrorKind) {
        self.add_error(error);
        self.eat();
    }

    pub(crate) fn eat_error_as(&mut self, kind: SyntaxKind, error: ParseErrorKind) {
        self.add_error(error);
        self.eat_as(kind);
    }

    pub(crate) fn add_error(&mut self, error: ParseErrorKind) {
        let span = self.lexer.span();
        self.errors.push(ParseError::new(
            TextRange::new(
                TextSize::from(span.start as u32),
                TextSize::from(span.end as u32),
            ),
            error,
        ));
    }

    pub(crate) fn start_node(&mut self, kind: SyntaxKind) {
        self.green.start_node(kind.into());
    }

    pub(crate) fn finish_node(&mut self) {
        self.green.finish_node();
    }

    pub(crate) fn finish(self) -> Parse {
        Parse {
            errors: self.errors,
            green: self.green.finish(),
        }
    }

    pub(crate) fn statement_closed(&self) -> bool {
        self.statement_closed
    }

    pub(crate) fn set_statement_closed(&mut self, v: bool) {
        self.statement_closed = v;
    }
}
