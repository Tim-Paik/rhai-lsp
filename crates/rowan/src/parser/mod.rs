//! This module contains all parsing-related tools including a
//! recursive-descent parser.

use crate::syntax::{SyntaxKind, SyntaxNode};
use rowan::GreenNode;
use thiserror::Error;

use self::context::Context;

mod context;
pub mod parsers;

/// A flexible parser.
/// 
/// Parsing happens via given parser functions (also found in [`parsers`]).
/// 
/// # Example Uses
/// 
/// ```
/// use rhai_rowan::parser::Parser;
/// // Parse some Rhai code (file).
/// let parser = Parser::new(r#"print("hello rhai!")"#);
/// let parse = parser.parse();
/// 
/// assert!(parse.errors.is_empty());
/// ```
/// 
/// ```
/// use rhai_rowan::parser::Parser;
/// use rhai_rowan::parser::parsers::parse_expr;
///
/// // Parse an expression explicitly.
/// let mut parser = Parser::new(r#"print("hello rhai!")"#);
/// parser.execute(parse_expr);
/// let parse = parser.finish();
/// 
/// assert!(parse.errors.is_empty());
/// ```
pub struct Parser<'src> {
    context: Context<'src>,
}

impl<'src> Parser<'src> {
    /// Create a new parser with the given source.
    #[must_use]
    pub fn new(source: &'src str) -> Self {
        Self {
            context: Context::new(source),
        }
    }

    /// Finish parsing.
    /// 
    /// # Panics
    /// 
    /// If no parser function was called,
    /// or a parser was left in an invalid state (bug).
    #[must_use]
    pub fn finish(self) -> Parse {
        self.context.finish()
    }
}

impl<'src> Parser<'src> {
    /// Execute a parser function for the given parser.
    pub fn execute<F: FnOnce(&mut Context)>(&mut self, f: F) {
        f(&mut self.context);
    }
}

/// The result of parsing, containing all errors
/// and the final green tree.
#[derive(Debug, Clone)]
pub struct Parse {
    /// Parse errors.
    pub errors: Vec<ParseError>,
    /// Parsed green tree.
    pub green: GreenNode,
}

impl Parse {
    /// Turn the result green tree into a CST.
    /// *This ignores errors*, the resulting tree
    /// can be potentially syntactically invalid.
    #[must_use]
    pub fn clone_syntax(&self) -> SyntaxNode {
        SyntaxNode::new_root(self.green.clone())
    }

    /// Turn the result green tree into a CST.
    /// *This ignores errors*, the resulting tree
    /// can be potentially syntactically invalid.
    #[must_use]
    pub fn into_syntax(self) -> SyntaxNode {
        SyntaxNode::new_root(self.green)
    }
}

/// A parse (syntax) error.
#[derive(Debug, Clone, PartialEq, Eq, Error)]
#[error("parse error at {range:?}: {kind}")]
pub struct ParseError {
    /// The span of the error in the parsed source.
    pub range: rowan::TextRange,
    /// Error kind.
    pub kind: ParseErrorKind,
}

impl ParseError {
    fn new(range: rowan::TextRange, kind: ParseErrorKind) -> Self {
        Self { range, kind }
    }
}

/// All the non-fatal parsing errors that can occur.
#[derive(Debug, PartialEq, Eq, Clone, Error)]
pub enum ParseErrorKind {
    #[error("unexpected EOF")]
    UnexpectedEof,

    #[error(r#"invalid input"#)]
    InvalidInput,

    #[error(r#"unexpected token"#)]
    UnexpectedToken,

    #[error(r#"expected token "{0:?}""#)]
    ExpectedToken(SyntaxKind),

    #[error(
        r#"expected one of "{}""#,
        .0
            .iter()
            .map(|s|format!("{:?}", s))
            .collect::<Vec<String>>().join(",")
    )]
    ExpectedOneOfTokens(Vec<SyntaxKind>),
}
