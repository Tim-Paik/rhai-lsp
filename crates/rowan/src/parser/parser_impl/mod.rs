#![deny(unreachable_patterns)]

use super::context::Context;
use crate::parser::ParseErrorKind;
use crate::syntax::SyntaxKind::*;
use crate::T;

macro_rules! require_token {
    ($ctx:ident) => {
        match $ctx.token() {
            Some(t) => t,
            None => return $ctx.eat_error(ParseErrorKind::UnexpectedEof),
        }
    };

    ($ctx:ident, finish_node) => {
        match $ctx.token() {
            Some(t) => t,
            None => {
                $ctx.finish_node();
                $ctx.eat_error(ParseErrorKind::UnexpectedEof);
                return;
            }
        }
    };
}

macro_rules! expect_token {
    ($ctx:ident, finish_node, $($token:tt)*) => {
        match $ctx.token() {
            Some(t) => match t {
                $($token)* => {
                    $ctx.eat();
                },
                _ => {
                    $ctx.finish_node();
                    $ctx.eat_error(ParseErrorKind::ExpectedToken($($token)*));
                    return;
                }
            }
            None => {
                $ctx.finish_node();
                $ctx.eat_error(ParseErrorKind::ExpectedToken($($token)*));
                return;
            }
        }
    };
    ($ctx:ident, $($token:tt)+) => {
        match $ctx.token() {
            Some(t) => match t {
                $($token)* => {
                    $ctx.eat();
                },
                _ => return $ctx.eat_error(ParseErrorKind::UnexpectedEof),
            }
            None => return $ctx.eat_error(ParseErrorKind::UnexpectedEof),
        }
    };

}

impl<'src> super::Parser<'src> {
    pub fn parse_file(&mut self) {
        parse_file(&mut self.context);
    }
}

fn parse_file(ctx: &mut Context) {
    ctx.start_node(FILE);
    if let Some(SHEBANG) = ctx.token() {
        parse_shebang(ctx);
    }

    ctx.set_statement_closed(true);
    while ctx.token().is_some() {
        if !ctx.statement_closed() {
            ctx.add_error(ParseErrorKind::ExpectedToken(T![";"]));
        }

        parse_stmt(ctx);
    }
    ctx.set_statement_closed(true);

    ctx.finish_node()
}

fn parse_shebang(ctx: &mut Context) {
    let token = require_token!(ctx);

    if !matches!(token, SHEBANG) {
        return ctx.eat_error(ParseErrorKind::ExpectedToken(SHEBANG));
    }

    ctx.eat()
}

fn parse_stmt(ctx: &mut Context) {
    let token = require_token!(ctx);
    ctx.start_node(STMT);
    ctx.set_statement_closed(false);

    if token == T![";"] {
        ctx.finish_node();
        ctx.set_statement_closed(true);
        return ctx.eat();
    }

    parse_item(ctx);

    if let Some(token) = ctx.token() {
        if token == T![";"] {
            ctx.eat();
            ctx.set_statement_closed(true);
        }
    }

    ctx.finish_node();
}

fn parse_item(ctx: &mut Context) {
    ctx.start_node(ITEM);

    // Parse doc comments if any.
    while matches!(
        require_token!(ctx, finish_node),
        COMMENT_BLOCK_DOC | COMMENT_LINE_DOC
    ) {
        ctx.eat();
    }

    parse_expr(ctx);

    ctx.finish_node();
}

fn parse_expr(ctx: &mut Context) {
    ctx.start_node(EXPR);
    let token = require_token!(ctx, finish_node);

    match token {
        T!["let"] => parse_expr_let(ctx),
        T!["const"] => parse_expr_const(ctx),
        T!["{"] => parse_expr_block(ctx),
        T!["fn"] => parse_expr_fn(ctx),
        LIT_INT | LIT_FLOAT | LIT_BOOL | LIT_STR | LIT_CHAR => parse_expr_lit(ctx),
        IDENT => parse_expr_ident(ctx),
        _ => ctx.eat_error(ParseErrorKind::UnexpectedToken),
    }

    ctx.finish_node();
}

fn parse_expr_ident(ctx: &mut Context) {
    ctx.start_node(EXPR_IDENT);
    parse_ident(ctx);
    ctx.finish_node();
}

fn parse_expr_lit(ctx: &mut Context) {
    ctx.start_node(EXPR_LIT);
    parse_lit(ctx);
    ctx.finish_node();
}

fn parse_expr_let(ctx: &mut Context) {
    ctx.start_node(EXPR_LET);

    expect_token!(ctx, finish_node, T!["let"]);
    expect_token!(ctx, finish_node, T!["ident"]);
    expect_token!(ctx, finish_node, T!["="]);

    parse_expr(ctx);

    ctx.finish_node();
}

fn parse_expr_const(ctx: &mut Context) {
    ctx.start_node(EXPR_CONST);

    expect_token!(ctx, finish_node, T!["const"]);
    expect_token!(ctx, finish_node, T!["ident"]);
    expect_token!(ctx, finish_node, T!["="]);

    parse_expr(ctx);

    ctx.finish_node();
}

fn parse_expr_block(ctx: &mut Context) {
    ctx.start_node(EXPR_BLOCK);

    expect_token!(ctx, finish_node, T!["{"]);

    ctx.set_statement_closed(true);
    loop {
        let token = require_token!(ctx, finish_node);

        if token == T!["}"] {
            break;
        }

        if !ctx.statement_closed() {
            ctx.add_error(ParseErrorKind::ExpectedToken(T![";"]));
        }

        parse_stmt(ctx);
    }

    expect_token!(ctx, finish_node, T!["}"]);

    // Blocks also don't require ";" when used as statements.
    ctx.set_statement_closed(true);

    ctx.finish_node();
}

fn parse_expr_fn(ctx: &mut Context) {
    ctx.start_node(EXPR_FN);

    expect_token!(ctx, finish_node, T!["fn"]);
    expect_token!(ctx, finish_node, T!["ident"]);

    parse_param_list(ctx);
    parse_expr_block(ctx);

    ctx.finish_node();
}

fn parse_param_list(ctx: &mut Context) {
    ctx.start_node(PARAM_LIST);

    let start_token = require_token!(ctx, finish_node);

    match start_token {
        T!["("] | T!["|"] => {
            ctx.eat();
        }
        _ => {
            ctx.eat_error(ParseErrorKind::ExpectedOneOfTokens(vec![T!["("], T!["|"]]));
            return;
        }
    }

    loop {
        let token = require_token!(ctx, finish_node);
        if matches!(
            (start_token, token),
            (T!["("], T![")"]) | (T!["|"], T!["|"])
        ) {
            ctx.eat();
            break;
        }

        parse_param(ctx);

        let end_token = require_token!(ctx, finish_node);

        match (start_token, end_token) {
            (T!["("], T![")"]) | (T!["|"], T!["|"]) => {
                ctx.eat();
                break;
            }
            (_, T![","]) => {
                ctx.eat();
            }
            _ => {
                ctx.eat_error(ParseErrorKind::ExpectedToken(T![","]));
                break;
            }
        }
    }

    ctx.finish_node();
}

fn parse_param(ctx: &mut Context) {
    ctx.start_node(PARAM);

    expect_token!(ctx, finish_node, T!["ident"]);

    ctx.finish_node();
}

fn parse_lit(ctx: &mut Context) {
    ctx.start_node(LIT);
    let token = require_token!(ctx, finish_node);

    match token {
        LIT_INT | LIT_FLOAT | LIT_BOOL | LIT_STR | LIT_CHAR => {
            ctx.eat();
        }
        _ => {
            ctx.eat_error(ParseErrorKind::ExpectedOneOfTokens(vec![
                LIT_INT, LIT_FLOAT, LIT_BOOL, LIT_STR, LIT_CHAR,
            ]));
        }
    }

    ctx.finish_node();
}

fn parse_ident(ctx: &mut Context) {
    let token = require_token!(ctx);

    match token {
        IDENT => ctx.eat(),
        _ => ctx.eat_error(ParseErrorKind::ExpectedToken(IDENT)),
    }
}

// Expr =
// | ExprUni
// | ExprBin
// | ExprParen
// | ExprArray
// | ExprIndex
// | ExprObject
// | ExprCall
// | ExprMethodCall
// | ExprAccess
// | ExprClosure
// | ExprIf
// | ExprLoop
// | ExprFor
// | ExprWhile
// | ExprBreak
// | ExprContinue
// | ExprSwitch
// | ExprReturn
// | ExprPath
// | ExprImport
