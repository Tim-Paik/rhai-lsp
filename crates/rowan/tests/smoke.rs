use rhai_rowan::parser::{Operator, Parser};
use test_case::test_case;

#[test_case("simple", include_str!("../../../testdata/valid/simple.rhai"))]
#[test_case("array", include_str!("../../../testdata/valid/array.rhai"))]
#[test_case("assignment", include_str!("../../../testdata/valid/assignment.rhai"))]
#[test_case("comments", include_str!("../../../testdata/valid/comments.rhai"))]
#[test_case("fibonacci", include_str!("../../../testdata/valid/fibonacci.rhai"))]
#[test_case("for1", include_str!("../../../testdata/valid/for1.rhai"))]
#[test_case("for2", include_str!("../../../testdata/valid/for2.rhai"))]
#[test_case("function_decl1", include_str!("../../../testdata/valid/function_decl1.rhai"))]
#[test_case("function_decl2", include_str!("../../../testdata/valid/function_decl2.rhai"))]
#[test_case("function_decl3", include_str!("../../../testdata/valid/function_decl3.rhai"))]
#[test_case("function_decl4", include_str!("../../../testdata/valid/function_decl4.rhai"))]
#[test_case("if1", include_str!("../../../testdata/valid/if1.rhai"))]
#[test_case("if2", include_str!("../../../testdata/valid/if2.rhai"))]
#[test_case("loop", include_str!("../../../testdata/valid/loop.rhai"))]
#[test_case("mat_mul", include_str!("../../../testdata/valid/mat_mul.rhai"))]
#[test_case("module", include_str!("../../../testdata/valid/module.rhai"))]
#[test_case("oop", include_str!("../../../testdata/valid/oop.rhai"))]
#[test_case("op1", include_str!("../../../testdata/valid/op1.rhai"))]
#[test_case("op2", include_str!("../../../testdata/valid/op2.rhai"))]
#[test_case("op3", include_str!("../../../testdata/valid/op3.rhai"))]
#[test_case("primes", include_str!("../../../testdata/valid/primes.rhai"))]
#[test_case("speed_test", include_str!("../../../testdata/valid/speed_test.rhai"))]
#[test_case("string", include_str!("../../../testdata/valid/string.rhai"))]
#[test_case("strings_map", include_str!("../../../testdata/valid/strings_map.rhai"))]
#[test_case("switch", include_str!("../../../testdata/valid/switch.rhai"))]
#[test_case("while", include_str!("../../../testdata/valid/while.rhai"))]
#[test_case("char", include_str!("../../../testdata/valid/char.rhai"))]
#[test_case("throw_try_catch", include_str!("../../../testdata/valid/throw_try_catch.rhai"))]
#[test_case("optional_ops", include_str!("../../../testdata/valid/optional_ops.rhai"))]
fn parse_valid(name: &str, src: &str) {
    let parse = Parser::new(src).parse_script();
    assert!(parse.errors.is_empty(), "{:#?}", parse.errors);

    let mut engine = rhai::Engine::new();
    engine.set_max_expr_depths(0, 0);

    if src.starts_with("#!") {
        engine
            .compile(
                src.lines()
                    .skip(1)
                    .map(ToString::to_string)
                    .collect::<Vec<String>>()
                    .join("\n"),
            )
            .unwrap();
    } else {
        engine.compile(src).unwrap();
    }

    insta::with_settings!(
        { snapshot_suffix => name },
        {
            insta::assert_snapshot!(format!("{:#?}", parse.into_syntax()));
        }
    );
}

#[test]
fn parse_custom_operator() {
    let src = include_str!("../../../testdata/valid/operators.rhai");

    let parse = Parser::new(src)
        .with_operator("over", Operator::default())
        .parse_script();
    assert!(parse.errors.is_empty(), "{:#?}", parse.errors);

    insta::assert_snapshot!(format!("{:#?}", parse.into_syntax()));

    let parse = Parser::new(src).parse_script();
    assert!(!parse.errors.is_empty());
}
