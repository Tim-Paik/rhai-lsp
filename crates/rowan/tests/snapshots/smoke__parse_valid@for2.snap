---
source: crates/rowan/tests/smoke.rs
assertion_line: 65
expression: "format!(\"{:#?}\", parse.into_syntax())"
---
RHAI@0..422
  COMMENT_LINE@0..29 "// This script runs f ..."
  WHITESPACE@29..31 "\n\n"
  STMT@31..53
    ITEM@31..52
      EXPR@31..52
        EXPR_CONST@31..52
          KW_CONST@31..36 "const"
          WHITESPACE@36..37 " "
          IDENT@37..40 "MAX"
          WHITESPACE@40..41 " "
          OP_ASSIGN@41..42 "="
          EXPR@42..52
            WHITESPACE@42..43 " "
            EXPR_LIT@43..52
              LIT@43..52
                LIT_INT@43..52 "1_000_000"
    PUNCT_SEMI@52..53 ";"
  WHITESPACE@53..55 "\n\n"
  STMT@55..104
    ITEM@55..103
      EXPR@55..103
        EXPR_CALL@55..103
          EXPR@55..60
            EXPR_IDENT@55..60
              IDENT@55..60 "print"
          ARG_LIST@60..103
            PUNCT_PAREN_START@60..61 "("
            EXPR@61..102
              EXPR_LIT@61..102
                LIT@61..102
                  LIT_STR_TEMPLATE@61..102
                    LIT_STR@61..86 "`Iterating an array w ..."
                    INTERPOLATION_START@86..88 "${"
                    LIT_STR_TEMPLATE_INTERPOLATION@88..91
                      STMT@88..91
                        ITEM@88..91
                          EXPR@88..91
                            EXPR_IDENT@88..91
                              IDENT@88..91 "MAX"
                    PUNCT_BRACE_END@91..92 "}"
                    LIT_STR@92..102 " items...`"
            PUNCT_PAREN_END@102..103 ")"
    PUNCT_SEMI@103..104 ";"
  WHITESPACE@104..106 "\n\n"
  STMT@106..128
    ITEM@106..127
      EXPR@106..127
        EXPR_CALL@106..127
          EXPR@106..111
            EXPR_IDENT@106..111
              IDENT@106..111 "print"
          ARG_LIST@111..127
            PUNCT_PAREN_START@111..112 "("
            EXPR@112..126
              EXPR_LIT@112..126
                LIT@112..126
                  LIT_STR@112..126 "\"Ready... Go!\""
            PUNCT_PAREN_END@126..127 ")"
    PUNCT_SEMI@127..128 ";"
  WHITESPACE@128..130 "\n\n"
  STMT@130..152
    ITEM@130..151
      EXPR@130..151
        EXPR_LET@130..151
          KW_LET@130..133 "let"
          WHITESPACE@133..134 " "
          IDENT@134..137 "now"
          WHITESPACE@137..138 " "
          OP_ASSIGN@138..139 "="
          EXPR@139..151
            EXPR_CALL@139..151
              EXPR@139..149
                WHITESPACE@139..140 " "
                EXPR_IDENT@140..149
                  IDENT@140..149 "timestamp"
              ARG_LIST@149..151
                PUNCT_PAREN_START@149..150 "("
                PUNCT_PAREN_END@150..151 ")"
    PUNCT_SEMI@151..152 ";"
  WHITESPACE@152..154 "\n\n"
  STMT@154..168
    ITEM@154..167
      EXPR@154..167
        EXPR_LET@154..167
          KW_LET@154..157 "let"
          WHITESPACE@157..158 " "
          IDENT@158..162 "list"
          WHITESPACE@162..163 " "
          OP_ASSIGN@163..164 "="
          EXPR@164..167
            WHITESPACE@164..165 " "
            EXPR_ARRAY@165..167
              PUNCT_BRACKET_START@165..166 "["
              PUNCT_BRACKET_END@166..167 "]"
    PUNCT_SEMI@167..168 ";"
  WHITESPACE@168..170 "\n\n"
  COMMENT_LINE@170..188 "// Loop over range"
  WHITESPACE@188..189 "\n"
  STMT@189..228
    ITEM@189..226
      EXPR@189..226
        EXPR_FOR@189..226
          KW_FOR@189..192 "for"
          PAT@192..194
            WHITESPACE@192..193 " "
            PAT_IDENT@193..194
              IDENT@193..194 "i"
          WHITESPACE@194..195 " "
          KW_IN@195..197 "in"
          EXPR@197..205
            EXPR_BINARY@197..205
              EXPR@197..199
                WHITESPACE@197..198 " "
                EXPR_LIT@198..199
                  LIT@198..199
                    LIT_INT@198..199 "0"
              OP_RANGE@199..201 ".."
              EXPR@201..205
                EXPR_IDENT@201..205
                  IDENT@201..204 "MAX"
                  WHITESPACE@204..205 " "
          EXPR_BLOCK@205..226
            PUNCT_BRACE_START@205..206 "{"
            WHITESPACE@206..211 "\n    "
            STMT@211..224
              ITEM@211..223
                EXPR@211..223
                  EXPR_BINARY@211..223
                    EXPR@211..215
                      EXPR_IDENT@211..215
                        IDENT@211..215 "list"
                    PUNCT_DOT@215..216 "."
                    EXPR@216..223
                      EXPR_CALL@216..223
                        EXPR@216..220
                          EXPR_IDENT@216..220
                            IDENT@216..220 "push"
                        ARG_LIST@220..223
                          PUNCT_PAREN_START@220..221 "("
                          EXPR@221..222
                            EXPR_IDENT@221..222
                              IDENT@221..222 "i"
                          PUNCT_PAREN_END@222..223 ")"
              PUNCT_SEMI@223..224 ";"
            WHITESPACE@224..225 "\n"
            PUNCT_BRACE_END@225..226 "}"
    WHITESPACE@226..228 "\n\n"
  STMT@228..270
    ITEM@228..269
      EXPR@228..269
        EXPR_CALL@228..269
          EXPR@228..233
            EXPR_IDENT@228..233
              IDENT@228..233 "print"
          ARG_LIST@233..269
            PUNCT_PAREN_START@233..234 "("
            EXPR@234..268
              EXPR_LIT@234..268
                LIT@234..268
                  LIT_STR_TEMPLATE@234..268
                    LIT_STR@234..242 "`Time = "
                    INTERPOLATION_START@242..244 "${"
                    LIT_STR_TEMPLATE_INTERPOLATION@244..255
                      STMT@244..255
                        ITEM@244..255
                          EXPR@244..255
                            EXPR_BINARY@244..255
                              EXPR@244..247
                                EXPR_IDENT@244..247
                                  IDENT@244..247 "now"
                              PUNCT_DOT@247..248 "."
                              EXPR@248..255
                                EXPR_IDENT@248..255
                                  IDENT@248..255 "elapsed"
                    PUNCT_BRACE_END@255..256 "}"
                    LIT_STR@256..268 " seconds...`"
            PUNCT_PAREN_END@268..269 ")"
    PUNCT_SEMI@269..270 ";"
  WHITESPACE@270..272 "\n\n"
  STMT@272..284
    ITEM@272..283
      EXPR@272..283
        EXPR_LET@272..283
          KW_LET@272..275 "let"
          WHITESPACE@275..276 " "
          IDENT@276..279 "sum"
          WHITESPACE@279..280 " "
          OP_ASSIGN@280..281 "="
          EXPR@281..283
            WHITESPACE@281..282 " "
            EXPR_LIT@282..283
              LIT@282..283
                LIT_INT@282..283 "0"
    PUNCT_SEMI@283..284 ";"
  WHITESPACE@284..286 "\n\n"
  COMMENT_LINE@286..304 "// Loop over array"
  WHITESPACE@304..305 "\n"
  STMT@305..338
    ITEM@305..336
      EXPR@305..336
        EXPR_FOR@305..336
          KW_FOR@305..308 "for"
          PAT@308..310
            WHITESPACE@308..309 " "
            PAT_IDENT@309..310
              IDENT@309..310 "i"
          WHITESPACE@310..311 " "
          KW_IN@311..313 "in"
          EXPR@313..319
            WHITESPACE@313..314 " "
            EXPR_IDENT@314..319
              IDENT@314..318 "list"
              WHITESPACE@318..319 " "
          EXPR_BLOCK@319..336
            PUNCT_BRACE_START@319..320 "{"
            WHITESPACE@320..325 "\n    "
            STMT@325..334
              ITEM@325..333
                EXPR@325..333
                  EXPR_BINARY@325..333
                    EXPR@325..329
                      EXPR_IDENT@325..329
                        IDENT@325..328 "sum"
                        WHITESPACE@328..329 " "
                    OP_ADD_ASSIGN@329..331 "+="
                    EXPR@331..333
                      WHITESPACE@331..332 " "
                      EXPR_IDENT@332..333
                        IDENT@332..333 "i"
              PUNCT_SEMI@333..334 ";"
            WHITESPACE@334..335 "\n"
            PUNCT_BRACE_END@335..336 "}"
    WHITESPACE@336..338 "\n\n"
  STMT@338..360
    ITEM@338..359
      EXPR@338..359
        EXPR_CALL@338..359
          EXPR@338..343
            EXPR_IDENT@338..343
              IDENT@338..343 "print"
          ARG_LIST@343..359
            PUNCT_PAREN_START@343..344 "("
            EXPR@344..358
              EXPR_LIT@344..358
                LIT@344..358
                  LIT_STR_TEMPLATE@344..358
                    LIT_STR@344..351 "`Sum = "
                    INTERPOLATION_START@351..353 "${"
                    LIT_STR_TEMPLATE_INTERPOLATION@353..356
                      STMT@353..356
                        ITEM@353..356
                          EXPR@353..356
                            EXPR_IDENT@353..356
                              IDENT@353..356 "sum"
                    PUNCT_BRACE_END@356..357 "}"
                    LIT_STR@357..358 "`"
            PUNCT_PAREN_END@358..359 ")"
    PUNCT_SEMI@359..360 ";"
  WHITESPACE@360..361 "\n"
  STMT@361..421
    ITEM@361..420
      EXPR@361..420
        EXPR_CALL@361..420
          EXPR@361..366
            EXPR_IDENT@361..366
              IDENT@361..366 "print"
          ARG_LIST@366..420
            PUNCT_PAREN_START@366..367 "("
            EXPR@367..419
              EXPR_LIT@367..419
                LIT@367..419
                  LIT_STR_TEMPLATE@367..419
                    LIT_STR@367..395 "`Finished. Total run  ..."
                    INTERPOLATION_START@395..397 "${"
                    LIT_STR_TEMPLATE_INTERPOLATION@397..408
                      STMT@397..408
                        ITEM@397..408
                          EXPR@397..408
                            EXPR_BINARY@397..408
                              EXPR@397..400
                                EXPR_IDENT@397..400
                                  IDENT@397..400 "now"
                              PUNCT_DOT@400..401 "."
                              EXPR@401..408
                                EXPR_IDENT@401..408
                                  IDENT@401..408 "elapsed"
                    PUNCT_BRACE_END@408..409 "}"
                    LIT_STR@409..419 " seconds.`"
            PUNCT_PAREN_END@419..420 ")"
    PUNCT_SEMI@420..421 ";"
  WHITESPACE@421..422 "\n"

